File: server-java-play-framework/app/controllers/RetailerApiController.java
package controllers;

import apimodels.AccountLoginResponse;
import java.math.BigDecimal;
import java.io.InputStream;
import apimodels.RetailerFullResponse;
import apimodels.RetailerResponse;
import apimodels.SirqulResponse;

import com.typesafe.config.Config;
import play.mvc.Controller;
import play.mvc.Result;
import play.mvc.Http;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;
import com.google.inject.Inject;
import java.io.File;
import play.libs.Files.TemporaryFile;
import openapitools.OpenAPIUtils;
import com.fasterxml.jackson.core.type.TypeReference;

import javax.validation.constraints.*;
import javax.validation.Valid;
import com.typesafe.config.Config;

import openapitools.OpenAPIUtils.ApiAction;

@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaPlayFrameworkCodegen", date = "2026-02-09T20:49:20.380387445Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
public class RetailerApiController extends Controller {
    private final RetailerApiControllerImpInterface imp;
    private final ObjectMapper mapper;
    private final Config configuration;

    @Inject
    private RetailerApiController(Config configuration, RetailerApiControllerImpInterface imp) {
        this.imp = imp;
        mapper = new ObjectMapper();
        this.configuration = configuration;
    }

    @ApiAction
    public Result createRetailer(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valuename = request.getQueryString("name");
        String name;
        if (valuename != null) {
            name = valuename;
        } else {
            throw new IllegalArgumentException("'name' parameter is required");
        }
        String valuestreetAddress = request.getQueryString("streetAddress");
        String streetAddress;
        if (valuestreetAddress != null) {
            streetAddress = valuestreetAddress;
        } else {
            streetAddress = null;
        }
        String valuestreetAddress2 = request.getQueryString("streetAddress2");
        String streetAddress2;
        if (valuestreetAddress2 != null) {
            streetAddress2 = valuestreetAddress2;
        } else {
            streetAddress2 = null;
        }
        String valuecity = request.getQueryString("city");
        String city;
        if (valuecity != null) {
            city = valuecity;
        } else {
            city = null;
        }
        String valuestate = request.getQueryString("state");
        String state;
        if (valuestate != null) {
            state = valuestate;
        } else {
            state = null;
        }
        String valuepostalCode = request.getQueryString("postalCode");
        String postalCode;
        if (valuepostalCode != null) {
            postalCode = valuepostalCode;
        } else {
            postalCode = null;
        }
        String valuecountry = request.getQueryString("country");
        String country;
        if (valuecountry != null) {
            country = valuecountry;
        } else {
            country = null;
        }
        String valuebusinessPhone = request.getQueryString("businessPhone");
        String businessPhone;
        if (valuebusinessPhone != null) {
            businessPhone = valuebusinessPhone;
        } else {
            businessPhone = null;
        }
        String valuebusinessPhoneExt = request.getQueryString("businessPhoneExt");
        String businessPhoneExt;
        if (valuebusinessPhoneExt != null) {
            businessPhoneExt = valuebusinessPhoneExt;
        } else {
            businessPhoneExt = null;
        }
        String valuewebsite = request.getQueryString("website");
        String website;
        if (valuewebsite != null) {
            website = valuewebsite;
        } else {
            website = null;
        }
        String valueemail = request.getQueryString("email");
        String email;
        if (valueemail != null) {
            email = valueemail;
        } else {
            email = null;
        }
        String valuefacebookUrl = request.getQueryString("facebookUrl");
        String facebookUrl;
        if (valuefacebookUrl != null) {
            facebookUrl = valuefacebookUrl;
        } else {
            facebookUrl = null;
        }
        String valuetwitterUrl = request.getQueryString("twitterUrl");
        String twitterUrl;
        if (valuetwitterUrl != null) {
            twitterUrl = valuetwitterUrl;
        } else {
            twitterUrl = null;
        }
        String valuelogo = request.getQueryString("logo");
        InputStream logo;
        if (valuelogo != null) {
            logo = valuelogo;
        } else {
            logo = null;
        }
        String valuelogoAssetId = request.getQueryString("logoAssetId");
        Long logoAssetId;
        if (valuelogoAssetId != null) {
            logoAssetId = Long.parseLong(valuelogoAssetId);
        } else {
            logoAssetId = null;
        }
        String valuepicture1 = request.getQueryString("picture1");
        InputStream picture1;
        if (valuepicture1 != null) {
            picture1 = valuepicture1;
        } else {
            picture1 = null;
        }
        String valuepicture1AssetId = request.getQueryString("picture1AssetId");
        Long picture1AssetId;
        if (valuepicture1AssetId != null) {
            picture1AssetId = Long.parseLong(valuepicture1AssetId);
        } else {
            picture1AssetId = null;
        }
        String valuepicture2 = request.getQueryString("picture2");
        InputStream picture2;
        if (valuepicture2 != null) {
            picture2 = valuepicture2;
        } else {
            picture2 = null;
        }
        String valuepicture2AssetId = request.getQueryString("picture2AssetId");
        Long picture2AssetId;
        if (valuepicture2AssetId != null) {
            picture2AssetId = Long.parseLong(valuepicture2AssetId);
        } else {
            picture2AssetId = null;
        }
        String valuecategoryIds = request.getQueryString("categoryIds");
        String categoryIds;
        if (valuecategoryIds != null) {
            categoryIds = valuecategoryIds;
        } else {
            categoryIds = null;
        }
        String valuecategoryIdsToAdd = request.getQueryString("categoryIdsToAdd");
        String categoryIdsToAdd;
        if (valuecategoryIdsToAdd != null) {
            categoryIdsToAdd = valuecategoryIdsToAdd;
        } else {
            categoryIdsToAdd = null;
        }
        String valuecategoryIdsToRemove = request.getQueryString("categoryIdsToRemove");
        String categoryIdsToRemove;
        if (valuecategoryIdsToRemove != null) {
            categoryIdsToRemove = valuecategoryIdsToRemove;
        } else {
            categoryIdsToRemove = null;
        }
        String valuefilterIds = request.getQueryString("filterIds");
        String filterIds;
        if (valuefilterIds != null) {
            filterIds = valuefilterIds;
        } else {
            filterIds = null;
        }
        String valuelatitude = request.getQueryString("latitude");
        Double latitude;
        if (valuelatitude != null) {
            latitude = Double.parseDouble(valuelatitude);
        } else {
            latitude = null;
        }
        String valuelongitude = request.getQueryString("longitude");
        Double longitude;
        if (valuelongitude != null) {
            longitude = Double.parseDouble(valuelongitude);
        } else {
            longitude = null;
        }
        String valuemetaData = request.getQueryString("metaData");
        String metaData;
        if (valuemetaData != null) {
            metaData = valuemetaData;
        } else {
            metaData = null;
        }
        String valuesearchTags = request.getQueryString("searchTags");
        String searchTags;
        if (valuesearchTags != null) {
            searchTags = valuesearchTags;
        } else {
            searchTags = null;
        }
        String valueretailerType = request.getQueryString("retailerType");
        String retailerType;
        if (valueretailerType != null) {
            retailerType = valueretailerType;
        } else {
            retailerType = null;
        }
        String valuevisibility = request.getQueryString("visibility");
        String visibility;
        if (valuevisibility != null) {
            visibility = valuevisibility;
        } else {
            visibility = null;
        }
        String valuecreateDefaultLocation = request.getQueryString("createDefaultLocation");
        Boolean createDefaultLocation;
        if (valuecreateDefaultLocation != null) {
            createDefaultLocation = Boolean.valueOf(valuecreateDefaultLocation);
        } else {
            createDefaultLocation = null;
        }
        String valueresponseFormat = request.getQueryString("responseFormat");
        String responseFormat;
        if (valueresponseFormat != null) {
            responseFormat = valueresponseFormat;
        } else {
            responseFormat = null;
        }
        return imp.createRetailerHttp(request, version, name, deviceId, accountId, streetAddress, streetAddress2, city, state, postalCode, country, businessPhone, businessPhoneExt, website, email, facebookUrl, twitterUrl, logo, logoAssetId, picture1, picture1AssetId, picture2, picture2AssetId, categoryIds, categoryIdsToAdd, categoryIdsToRemove, filterIds, latitude, longitude, metaData, searchTags, retailerType, visibility, createDefaultLocation, responseFormat);
    }

    @ApiAction
    public Result deleteRetailer(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valueretailerId = request.getQueryString("retailerId");
        Long retailerId;
        if (valueretailerId != null) {
            retailerId = Long.parseLong(valueretailerId);
        } else {
            retailerId = null;
        }
        return imp.deleteRetailerHttp(request, version, deviceId, accountId, retailerId);
    }

    @ApiAction
    public Result getRetailer(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valueretailerId = request.getQueryString("retailerId");
        Long retailerId;
        if (valueretailerId != null) {
            retailerId = Long.parseLong(valueretailerId);
        } else {
            throw new IllegalArgumentException("'retailerId' parameter is required");
        }
        String valueincludeCounts = request.getQueryString("includeCounts");
        Boolean includeCounts;
        if (valueincludeCounts != null) {
            includeCounts = Boolean.valueOf(valueincludeCounts);
        } else {
            includeCounts = null;
        }
        return imp.getRetailerHttp(request, version, retailerId, deviceId, accountId, includeCounts);
    }

    @ApiAction
    public Result getRetailers(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valueq = request.getQueryString("q");
        String q;
        if (valueq != null) {
            q = valueq;
        } else {
            q = null;
        }
        String valuekeyword = request.getQueryString("keyword");
        String keyword;
        if (valuekeyword != null) {
            keyword = valuekeyword;
        } else {
            keyword = null;
        }
        String valuecategoryIds = request.getQueryString("categoryIds");
        String categoryIds;
        if (valuecategoryIds != null) {
            categoryIds = valuecategoryIds;
        } else {
            categoryIds = null;
        }
        String valuefilterIds = request.getQueryString("filterIds");
        String filterIds;
        if (valuefilterIds != null) {
            filterIds = valuefilterIds;
        } else {
            filterIds = null;
        }
        String valuevisibility = request.getQueryString("visibility");
        String visibility;
        if (valuevisibility != null) {
            visibility = valuevisibility;
        } else {
            throw new IllegalArgumentException("'visibility' parameter is required");
        }
        String valuesortField = request.getQueryString("sortField");
        String sortField;
        if (valuesortField != null) {
            sortField = valuesortField;
        } else {
            throw new IllegalArgumentException("'sortField' parameter is required");
        }
        String valuedescending = request.getQueryString("descending");
        Boolean descending;
        if (valuedescending != null) {
            descending = Boolean.valueOf(valuedescending);
        } else {
            throw new IllegalArgumentException("'descending' parameter is required");
        }
        String valuei = request.getQueryString("_i");
        Integer i;
        if (valuei != null) {
            i = Integer.parseInt(valuei);
        } else {
            i = null;
        }
        String valuestart = request.getQueryString("start");
        Integer start;
        if (valuestart != null) {
            start = Integer.parseInt(valuestart);
        } else {
            throw new IllegalArgumentException("'start' parameter is required");
        }
        String valuel = request.getQueryString("_l");
        Integer l;
        if (valuel != null) {
            l = Integer.parseInt(valuel);
        } else {
            l = null;
        }
        String valuelimit = request.getQueryString("limit");
        Integer limit;
        if (valuelimit != null) {
            limit = Integer.parseInt(valuelimit);
        } else {
            throw new IllegalArgumentException("'limit' parameter is required");
        }
        String valueactiveOnly = request.getQueryString("activeOnly");
        Boolean activeOnly;
        if (valueactiveOnly != null) {
            activeOnly = Boolean.valueOf(valueactiveOnly);
        } else {
            throw new IllegalArgumentException("'activeOnly' parameter is required");
        }
        return imp.getRetailersHttp(request, version, visibility, sortField, descending, start, limit, activeOnly, deviceId, accountId, q, keyword, categoryIds, filterIds, i, l);
    }

    @ApiAction
    public Result retailerLoginCheck(Http.Request request, BigDecimal version) throws Exception {
        String valueusername = request.getQueryString("username");
        String username;
        if (valueusername != null) {
            username = valueusername;
        } else {
            throw new IllegalArgumentException("'username' parameter is required");
        }
        String valuepassword = request.getQueryString("password");
        String password;
        if (valuepassword != null) {
            password = valuepassword;
        } else {
            throw new IllegalArgumentException("'password' parameter is required");
        }
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valuelatitude = request.getQueryString("latitude");
        Double latitude;
        if (valuelatitude != null) {
            latitude = Double.parseDouble(valuelatitude);
        } else {
            latitude = null;
        }
        String valuelongitude = request.getQueryString("longitude");
        Double longitude;
        if (valuelongitude != null) {
            longitude = Double.parseDouble(valuelongitude);
        } else {
            longitude = null;
        }
        String valueappKey = request.getQueryString("appKey");
        String appKey;
        if (valueappKey != null) {
            appKey = valueappKey;
        } else {
            appKey = null;
        }
        return imp.retailerLoginCheckHttp(request, version, username, password, deviceId, latitude, longitude, appKey);
    }

    @ApiAction
    public Result updateRetailer(Http.Request request, BigDecimal version) throws Exception {
        String valuedeviceId = request.getQueryString("deviceId");
        String deviceId;
        if (valuedeviceId != null) {
            deviceId = valuedeviceId;
        } else {
            deviceId = null;
        }
        String valueaccountId = request.getQueryString("accountId");
        Long accountId;
        if (valueaccountId != null) {
            accountId = Long.parseLong(valueaccountId);
        } else {
            accountId = null;
        }
        String valueretailerId = request.getQueryString("retailerId");
        Long retailerId;
        if (valueretailerId != null) {
            retailerId = Long.parseLong(valueretailerId);
        } else {
            throw new IllegalArgumentException("'retailerId' parameter is required");
        }
        String valuename = request.getQueryString("name");
        String name;
        if (valuename != null) {
            name = valuename;
        } else {
            name = null;
        }
        String valuestreetAddress = request.getQueryString("streetAddress");
        String streetAddress;
        if (valuestreetAddress != null) {
            streetAddress = valuestreetAddress;
        } else {
            streetAddress = null;
        }
        String valuestreetAddress2 = request.getQueryString("streetAddress2");
        String streetAddress2;
        if (valuestreetAddress2 != null) {
            streetAddress2 = valuestreetAddress2;
        } else {
            streetAddress2 = null;
        }
        String valuecity = request.getQueryString("city");
        String city;
        if (valuecity != null) {
            city = valuecity;
        } else {
            city = null;
        }
        String valuestate = request.getQueryString("state");
        String state;
        if (valuestate != null) {
            state = valuestate;
        } else {
            state = null;
        }
        String valuepostalCode = request.getQueryString("postalCode");
        String postalCode;
        if (valuepostalCode != null) {
            postalCode = valuepostalCode;
        } else {
            postalCode = null;
        }
        String valuecountry = request.getQueryString("country");
        String country;
        if (valuecountry != null) {
            country = valuecountry;
        } else {
            country = null;
        }
        String valuebusinessPhone = request.getQueryString("businessPhone");
        String businessPhone;
        if (valuebusinessPhone != null) {
            businessPhone = valuebusinessPhone;
        } else {
            businessPhone = null;
        }
        String valuebusinessPhoneExt = request.getQueryString("businessPhoneExt");
        String businessPhoneExt;
        if (valuebusinessPhoneExt != null) {
            businessPhoneExt = valuebusinessPhoneExt;
        } else {
            businessPhoneExt = null;
        }
        String valuewebsite = request.getQueryString("website");
        String website;
        if (valuewebsite != null) {
            website = valuewebsite;
        } else {
            website = null;
        }
        String valueemail = request.getQueryString("email");
        String email;
        if (valueemail != null) {
            email = valueemail;
        } else {
            email = null;
        }
        String valuefacebookUrl = request.getQueryString("facebookUrl");
        String facebookUrl;
        if (valuefacebookUrl != null) {
            facebookUrl = valuefacebookUrl;
        } else {
            facebookUrl = null;
        }
        String valuetwitterUrl = request.getQueryString("twitterUrl");
        String twitterUrl;
        if (valuetwitterUrl != null) {
            twitterUrl = valuetwitterUrl;
        } else {
            twitterUrl = null;
        }
        String valuelogo = request.getQueryString("logo");
        InputStream logo;
        if (valuelogo != null) {
            logo = valuelogo;
        } else {
            logo = null;
        }
        String valuelogoAssetId = request.getQueryString("logoAssetId");
        Long logoAssetId;
        if (valuelogoAssetId != null) {
            logoAssetId = Long.parseLong(valuelogoAssetId);
        } else {
            logoAssetId = null;
        }
        String valuepicture1 = request.getQueryString("picture1");
        InputStream picture1;
        if (valuepicture1 != null) {
            picture1 = valuepicture1;
        } else {
            picture1 = null;
        }
        String valuepicture1AssetId = request.getQueryString("picture1AssetId");
        Long picture1AssetId;
        if (valuepicture1AssetId != null) {
            picture1AssetId = Long.parseLong(valuepicture1AssetId);
        } else {
            picture1AssetId = null;
        }
        String valuepicture2 = request.getQueryString("picture2");
        InputStream picture2;
        if (valuepicture2 != null) {
            picture2 = valuepicture2;
        } else {
            picture2 = null;
        }
        String valuepicture2AssetId = request.getQueryString("picture2AssetId");
        Long picture2AssetId;
        if (valuepicture2AssetId != null) {
            picture2AssetId = Long.parseLong(valuepicture2AssetId);
        } else {
            picture2AssetId = null;
        }
        String valuecategoryIds = request.getQueryString("categoryIds");
        String categoryIds;
        if (valuecategoryIds != null) {
            categoryIds = valuecategoryIds;
        } else {
            categoryIds = null;
        }
        String valuefilterIds = request.getQueryString("filterIds");
        String filterIds;
        if (valuefilterIds != null) {
            filterIds = valuefilterIds;
        } else {
            filterIds = null;
        }
        String valuelatitude = request.getQueryString("latitude");
        Double latitude;
        if (valuelatitude != null) {
            latitude = Double.parseDouble(valuelatitude);
        } else {
            latitude = null;
        }
        String valuelongitude = request.getQueryString("longitude");
        Double longitude;
        if (valuelongitude != null) {
            longitude = Double.parseDouble(valuelongitude);
        } else {
            longitude = null;
        }
        String valuemetaData = request.getQueryString("metaData");
        String metaData;
        if (valuemetaData != null) {
            metaData = valuemetaData;
        } else {
            metaData = null;
        }
        String valuesearchTags = request.getQueryString("searchTags");
        String searchTags;
        if (valuesearchTags != null) {
            searchTags = valuesearchTags;
        } else {
            searchTags = null;
        }
        String valueretailerType = request.getQueryString("retailerType");
        String retailerType;
        if (valueretailerType != null) {
            retailerType = valueretailerType;
        } else {
            retailerType = null;
        }
        String valuevisibility = request.getQueryString("visibility");
        String visibility;
        if (valuevisibility != null) {
            visibility = valuevisibility;
        } else {
            visibility = null;
        }
        String valueactive = request.getQueryString("active");
        Boolean active;
        if (valueactive != null) {
            active = Boolean.valueOf(valueactive);
        } else {
            active = null;
        }
        String valueresponseFormat = request.getQueryString("responseFormat");
        String responseFormat;
        if (valueresponseFormat != null) {
            responseFormat = valueresponseFormat;
        } else {
            responseFormat = null;
        }
        return imp.updateRetailerHttp(request, version, retailerId, deviceId, accountId, name, streetAddress, streetAddress2, city, state, postalCode, country, businessPhone, businessPhoneExt, website, email, facebookUrl, twitterUrl, logo, logoAssetId, picture1, picture1AssetId, picture2, picture2AssetId, categoryIds, filterIds, latitude, longitude, metaData, searchTags, retailerType, visibility, active, responseFormat);
    }

}

File: server-scala-play-server/app/api/RetailerApiController.scala
package api

import org.openapitools.OpenApiExceptions
import javax.inject.{Inject, Singleton}
import play.api.libs.json._
import play.api.mvc._
import model.AccountLoginResponse
import model.BigDecimal
import model.RetailerFullResponse
import model.RetailerResponse
import model.SirqulResponse
import play.api.libs.Files.TemporaryFile

@javax.annotation.Generated(value = Array("org.openapitools.codegen.languages.ScalaPlayFrameworkServerCodegen"), date = "2026-02-09T20:57:39.376804970Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
@Singleton
class RetailerApiController @Inject()(cc: ControllerComponents, api: RetailerApi) extends AbstractController(cc) {
  /**
    * POST /api/:version/retailer/create?deviceId=[value]&accountId=[value]&name=[value]&streetAddress=[value]&streetAddress2=[value]&city=[value]&state=[value]&postalCode=[value]&country=[value]&businessPhone=[value]&businessPhoneExt=[value]&website=[value]&email=[value]&facebookUrl=[value]&twitterUrl=[value]&logo=[value]&logoAssetId=[value]&picture1=[value]&picture1AssetId=[value]&picture2=[value]&picture2AssetId=[value]&categoryIds=[value]&categoryIdsToAdd=[value]&categoryIdsToRemove=[value]&filterIds=[value]&latitude=[value]&longitude=[value]&metaData=[value]&searchTags=[value]&retailerType=[value]&visibility=[value]&createDefaultLocation=[value]&responseFormat=[value]
    */
  def createRetailer(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): RetailerFullResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val name = request.getQueryString("name")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("name", "query string")
        }
        
      val streetAddress = request.getQueryString("streetAddress")
        
      val streetAddress2 = request.getQueryString("streetAddress2")
        
      val city = request.getQueryString("city")
        
      val state = request.getQueryString("state")
        
      val postalCode = request.getQueryString("postalCode")
        
      val country = request.getQueryString("country")
        
      val businessPhone = request.getQueryString("businessPhone")
        
      val businessPhoneExt = request.getQueryString("businessPhoneExt")
        
      val website = request.getQueryString("website")
        
      val email = request.getQueryString("email")
        
      val facebookUrl = request.getQueryString("facebookUrl")
        
      val twitterUrl = request.getQueryString("twitterUrl")
        
      val logo = request.getQueryString("logo")
        
      val logoAssetId = request.getQueryString("logoAssetId")
        .map(value => value.toLong)
        
      val picture1 = request.getQueryString("picture1")
        
      val picture1AssetId = request.getQueryString("picture1AssetId")
        .map(value => value.toLong)
        
      val picture2 = request.getQueryString("picture2")
        
      val picture2AssetId = request.getQueryString("picture2AssetId")
        .map(value => value.toLong)
        
      val categoryIds = request.getQueryString("categoryIds")
        
      val categoryIdsToAdd = request.getQueryString("categoryIdsToAdd")
        
      val categoryIdsToRemove = request.getQueryString("categoryIdsToRemove")
        
      val filterIds = request.getQueryString("filterIds")
        
      val latitude = request.getQueryString("latitude")
        .map(value => value.toDouble)
        
      val longitude = request.getQueryString("longitude")
        .map(value => value.toDouble)
        
      val metaData = request.getQueryString("metaData")
        
      val searchTags = request.getQueryString("searchTags")
        
      val retailerType = request.getQueryString("retailerType")
        
      val visibility = request.getQueryString("visibility")
        
      val createDefaultLocation = request.getQueryString("createDefaultLocation")
        .map(value => value.toBoolean)
        
      val responseFormat = request.getQueryString("responseFormat")
        
      api.createRetailer(version, name, deviceId, accountId, streetAddress, streetAddress2, city, state, postalCode, country, businessPhone, businessPhoneExt, website, email, facebookUrl, twitterUrl, logo, logoAssetId, picture1, picture1AssetId, picture2, picture2AssetId, categoryIds, categoryIdsToAdd, categoryIdsToRemove, filterIds, latitude, longitude, metaData, searchTags, retailerType, visibility, createDefaultLocation, responseFormat)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * POST /api/:version/retailer/delete?deviceId=[value]&accountId=[value]&retailerId=[value]
    */
  def deleteRetailer(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): SirqulResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val retailerId = request.getQueryString("retailerId")
        .map(value => value.toLong)
        
      api.deleteRetailer(version, deviceId, accountId, retailerId)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/:version/retailer/get?deviceId=[value]&accountId=[value]&retailerId=[value]&includeCounts=[value]
    */
  def getRetailer(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): RetailerFullResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val retailerId = request.getQueryString("retailerId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("retailerId", "query string")
        }
        
      val includeCounts = request.getQueryString("includeCounts")
        .map(value => value.toBoolean)
        
      api.getRetailer(version, retailerId, deviceId, accountId, includeCounts)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * GET /api/:version/retailer/search?deviceId=[value]&accountId=[value]&q=[value]&keyword=[value]&categoryIds=[value]&filterIds=[value]&visibility=[value]&sortField=[value]&descending=[value]&i=[value]&start=[value]&l=[value]&limit=[value]&activeOnly=[value]
    */
  def getRetailers(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): List[RetailerResponse] = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val q = request.getQueryString("q")
        
      val keyword = request.getQueryString("keyword")
        
      val categoryIds = request.getQueryString("categoryIds")
        
      val filterIds = request.getQueryString("filterIds")
        
      val visibility = request.getQueryString("visibility")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("visibility", "query string")
        }
        
      val sortField = request.getQueryString("sortField")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("sortField", "query string")
        }
        
      val descending = request.getQueryString("descending")
        .map(value => value.toBoolean)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("descending", "query string")
        }
        
      val i = request.getQueryString("_i")
        .map(value => value.toInt)
        
      val start = request.getQueryString("start")
        .map(value => value.toInt)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("start", "query string")
        }
        
      val l = request.getQueryString("_l")
        .map(value => value.toInt)
        
      val limit = request.getQueryString("limit")
        .map(value => value.toInt)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("limit", "query string")
        }
        
      val activeOnly = request.getQueryString("activeOnly")
        .map(value => value.toBoolean)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("activeOnly", "query string")
        }
        
      api.getRetailers(version, visibility, sortField, descending, start, limit, activeOnly, deviceId, accountId, q, keyword, categoryIds, filterIds, i, l)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * POST /api/:version/retailer/login?username=[value]&password=[value]&deviceId=[value]&latitude=[value]&longitude=[value]&appKey=[value]
    */
  def retailerLoginCheck(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): AccountLoginResponse = {
      val username = request.getQueryString("username")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("username", "query string")
        }
        
      val password = request.getQueryString("password")
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("password", "query string")
        }
        
      val deviceId = request.getQueryString("deviceId")
        
      val latitude = request.getQueryString("latitude")
        .map(value => value.toDouble)
        
      val longitude = request.getQueryString("longitude")
        .map(value => value.toDouble)
        
      val appKey = request.getQueryString("appKey")
        
      api.retailerLoginCheck(version, username, password, deviceId, latitude, longitude, appKey)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  /**
    * POST /api/:version/retailer/update?deviceId=[value]&accountId=[value]&retailerId=[value]&name=[value]&streetAddress=[value]&streetAddress2=[value]&city=[value]&state=[value]&postalCode=[value]&country=[value]&businessPhone=[value]&businessPhoneExt=[value]&website=[value]&email=[value]&facebookUrl=[value]&twitterUrl=[value]&logo=[value]&logoAssetId=[value]&picture1=[value]&picture1AssetId=[value]&picture2=[value]&picture2AssetId=[value]&categoryIds=[value]&filterIds=[value]&latitude=[value]&longitude=[value]&metaData=[value]&searchTags=[value]&retailerType=[value]&visibility=[value]&active=[value]&responseFormat=[value]
    */
  def updateRetailer(version: BigDecimal): Action[AnyContent] = Action { request =>
    def executeApi(): RetailerFullResponse = {
      val deviceId = request.getQueryString("deviceId")
        
      val accountId = request.getQueryString("accountId")
        .map(value => value.toLong)
        
      val retailerId = request.getQueryString("retailerId")
        .map(value => value.toLong)
        .getOrElse {
          throw new OpenApiExceptions.MissingRequiredParameterException("retailerId", "query string")
        }
        
      val name = request.getQueryString("name")
        
      val streetAddress = request.getQueryString("streetAddress")
        
      val streetAddress2 = request.getQueryString("streetAddress2")
        
      val city = request.getQueryString("city")
        
      val state = request.getQueryString("state")
        
      val postalCode = request.getQueryString("postalCode")
        
      val country = request.getQueryString("country")
        
      val businessPhone = request.getQueryString("businessPhone")
        
      val businessPhoneExt = request.getQueryString("businessPhoneExt")
        
      val website = request.getQueryString("website")
        
      val email = request.getQueryString("email")
        
      val facebookUrl = request.getQueryString("facebookUrl")
        
      val twitterUrl = request.getQueryString("twitterUrl")
        
      val logo = request.getQueryString("logo")
        
      val logoAssetId = request.getQueryString("logoAssetId")
        .map(value => value.toLong)
        
      val picture1 = request.getQueryString("picture1")
        
      val picture1AssetId = request.getQueryString("picture1AssetId")
        .map(value => value.toLong)
        
      val picture2 = request.getQueryString("picture2")
        
      val picture2AssetId = request.getQueryString("picture2AssetId")
        .map(value => value.toLong)
        
      val categoryIds = request.getQueryString("categoryIds")
        
      val filterIds = request.getQueryString("filterIds")
        
      val latitude = request.getQueryString("latitude")
        .map(value => value.toDouble)
        
      val longitude = request.getQueryString("longitude")
        .map(value => value.toDouble)
        
      val metaData = request.getQueryString("metaData")
        
      val searchTags = request.getQueryString("searchTags")
        
      val retailerType = request.getQueryString("retailerType")
        
      val visibility = request.getQueryString("visibility")
        
      val active = request.getQueryString("active")
        .map(value => value.toBoolean)
        
      val responseFormat = request.getQueryString("responseFormat")
        
      api.updateRetailer(version, retailerId, deviceId, accountId, name, streetAddress, streetAddress2, city, state, postalCode, country, businessPhone, businessPhoneExt, website, email, facebookUrl, twitterUrl, logo, logoAssetId, picture1, picture1AssetId, picture2, picture2AssetId, categoryIds, filterIds, latitude, longitude, metaData, searchTags, retailerType, visibility, active, responseFormat)
    }

    val result = executeApi()
    val json = Json.toJson(result)
    Ok(json)
  }

  private def splitCollectionParam(paramValues: String, collectionFormat: String): List[String] = {
    val splitBy =
      collectionFormat match {
        case "csv" => ",+"
        case "tsv" => "\t+"
        case "ssv" => " +"
        case "pipes" => "|+"
      }

    paramValues.split(splitBy).toList
  }
}
